Implement Apple TV Platform

fuzzard opened the pull request:
## Description
Implement the AppleTV platform for ATV4/4k

## Motivation and Context
Well here we are, years in the making. May as well say thanks to everyone to get it this far, Memphiz, Davilla (Team Mrmc), Sylvain, Kambala, Phunkyfish, Pogar, and anyone else who has helped that ive missed.

Notes:
PR is not in buildable individual commits. We have broken it up like this for, hopefully, for easier review, particularly the non platform code that is touched. Before Merge, the intent is to squash most into a single commit.
Feel free to update Labels as you see relevant.

Todo:
Currently a lot of refactoring work is being done around darwin_embedded platform as a whole. This will obviously bleed into tvos, but will tackle it later. Feel free to comment on any architecture related issues, but keep in mind, that will most likely go on the todo list for now.

## How Has This Been Tested?
ATV dev's, and some external users who have helped with testing

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [X] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

sy6sy2 committed:
[tvOS] Initial Platform Commit for Apple TVOS

Initial commit for TVOS platform.

fuzzard committed:
[tvOS] add binary-addons to xcode project

binary-addon building added to xcode project for tvOS.
The default behaviour is to add all addons to the project, but not add as a dependency.
This means the addons will NOT be built be default.
Documentation updated and shows how to selectively build addons if required and how to
add the auto building dependecy to the Xcode project.

fuzzard committed:
[ios-common] relocate Darwinnsuserdefaults to ios-common

Currently no use outside of tvos. Can be used for ios, so will put in ios-common

Rechi reviewed (changes requested):
You are not following the [code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

fuzzard committed:
[ios/tvos] Relocate CPUInfo for DarwinEmbed platforms

fuzzard commented:
Should be better now rechi. The 2 files that don't conform are 3rd party code we use for a gzip library

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -197,6 +197,7 @@
      if (pid == 0) {
          /* Child process */
          /*
+
```
Why is this line needed?

fuzzard replied:
was done by the diff command creating the patch

Rechi replied:
It for sure wasn't, the above hunk applies cleanly without this additional line.

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -214,7 +215,7 @@
 --- a/Modules/posixmodule.c
 +++ b/Modules/posixmodule.c
-@@ -197,6 +197,20 @@
+@@ -194,6 +194,20 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes, all ranges are correct. They are based off the preceding patches having been applied (5 patches) as laid out in the makefile.

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -231,11 +232,11 @@
 +#  undef HAVE_WAIT3
 +#  undef HAVE_WAIT4
-+#  undef HAVE_WAITPID
++//#  undef HAVE_WAITPID
```
Why patch commented `#  undef HAVE_WAITPID` in, instead of not patching this line in?

fuzzard replied:
at the time was just easier to update the patch by modifying the comment rather than rediff it all. I'll get it removed proper now

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -231,11 +232,11 @@
  /*[clinic input]
  # one of the few times we lie about this name!
-@@ -475,6 +489,10 @@
+@@ -472,6 +486,10 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -246,7 +247,7 @@
  void
  PyOS_AfterFork(void)
-@@ -5040,6 +5058,8 @@
+@@ -5037,6 +5055,8 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -255,7 +256,7 @@
  
      free_string_array(argvlist, argc);
-@@ -5116,6 +5136,12 @@
+@@ -5113,6 +5133,12 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -268,7 +269,7 @@
  
      posix_path_error(path);
-@@ -5416,7 +5442,7 @@
+@@ -5413,7 +5439,7 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes

Rechi commented on tools/depends/target/python3/darwin_embedded.patch:
```diff
@@ -277,7 +278,7 @@
          /* child: this clobbers and resets the import lock. */
          PyOS_AfterFork_Child();
-@@ -5447,7 +5473,7 @@
+@@ -5444,7 +5470,7 @@
```
Is the current range information incorrect?

fuzzard replied:
Yes

Rechi commented on cmake/platform/darwin_embedded/tvos.cmake:
```diff
@@ -0,0 +1,3 @@
+set(PLATFORM_REQUIRED_DEPS OpenGLES)
+set(APP_RENDER_SYSTEM gles)
+list(APPEND PLATFORM_DEFINES -DTARGET_DARWIN_TVOS)
```
add an empty line at the end of file

Rechi commented on tools/buildsteps/tvos/prepare-xbmc:
```diff
@@ -4,6 +4,6 @@ XBMC_PLATFORM_DIR=tvos
 #build binary addons before building xbmc...
 #make sure that binary_addons don't clean the native tools
-#here
+#here (e.x. on release builds where pathChanged always returns 1
```
really?

Rechi commented on xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/Contents.json:
```diff
@@ -0,0 +1,32 @@
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
add an empty line at the end of file

Rechi commented on xbmc/platform/darwin/tvos/Assets.xcassets/Assets.brandassets/topshelf.imageset/Contents.json:
```diff
@@ -0,0 +1,16 @@
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
add an empty line at the end of file

Rechi commented on xbmc/settings/DisplaySettings.cpp:
```diff
@@ -822,19 +831,14 @@ void CDisplaySettings::SettingOptionsPreferredStereoscopicViewModesFiller(Settin
-  CWinSystemOSX *winSystem = dynamic_cast<CWinSystemOSX*>(CServiceBroker::GetWinSystem());
-#elif defined(TARGET_DARWIN_IOS)
-  CWinSystemIOS *winSystem = dynamic_cast<CWinSystemIOS*>(CServiceBroker::GetWinSystem());
-#endif
```
The refactoring should be moved into a separate commit.

fuzzard replied:
Done.

Rechi commented on xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.h:
```diff
@@ -1,28 +1,50 @@
+ *  You should have received a copy of the GNU General Public License
+ *  along with MrMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
```
Why?

fuzzard replied:
Was a backport from mrmc codebase, copied the licensing as is from that

Rechi replied:
Please use SPDX licensing header.

Rechi commented on xbmc/platform/darwin/ios-common/DarwinNSUserDefaults.mm:
```diff
@@ -2,27 +2,70 @@
+ *  You should have received a copy of the GNU General Public License
+ *  along with MrMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
```
Why?

Rechi commented on cmake/scripts/darwin_embedded/Install.cmake:
```diff
@@ -80,6 +80,21 @@ set_target_properties(${APP_NAME_LC} PROPERTIES XCODE_ATTRIBUTE_CODE_SIGN_IDENTI
+    add_dependencies(${APP_NAME_LC} binary-addons)
+  endif()
+  unset(_addons)
+endif()
```
Why only add this for tvOS?

fuzzard replied:
Was put in the cmakebuildsys PR earlier, but was asked to be pulled out for "discussion".
Because of timelines, i felt it easier to attribute to tvos only, and if its desired for ios, i can remove the platform check at a later date.

Rechi replied:
It wasn't related to the previous PR and still isn't to this one. If you want this feature, open a separate PR, which implements it for all platforms.

fuzzard replied:
Its a feature to ease building for the apple tv platform, how is that not relevant here? Just because it could be used for ios is not the point. We are working towards updating ios stuff as well, and thats all still to come, why not add a feature for THIS platform now?

Rechi replied:
It's not a feature for tvOS only and also not for tvOS and iOS only, it is a general feature.
The reason I ask for doing it right from the beginning is, that otherwise it might get left behind.

fuzzard replied:
Honestly i dont have time to vet it for macos right now. I know it works for ios, but again, i PR'd this early knowing there would be feedback, and i understand where you are coming from regarding "forgotten" catch up features for other platforms, but the point is to get this in for Alpha no? Not everything will be perfect, and theres plenty of todo's. 

I would rather put it through in its known tested state, and when time permits and proper testing is done, open up the platform check to others.

Rechi reviewed (commented):
For the most commits in this PR I don't know what changes to expect as it is the directory path of the changed files.

fuzzard committed:
[settings] SettingOptionsMonitorsFiller refactor

implement WIN_SYSTEM_CLASS define to refactor SettingOptionsMonitorsFiller winSystem
object creation. Less duplicate code.

fuzzard commented:
> For the most commits in this PR I don't know what changes to expect as it is the directory path of the changed files.

Not sure what you mean by this? is this for the cmake treedata split for tvos/ios

Rechi commented:
E.g. the commit message `[tvOS] system`, `[tvOS] xbmc` or `[tvOS] xbmc/addons`. I have no idea what it changes, without looking at it.
Actually I see no reason to split the changes for tvOS per xbmc subdirectory, because they do not work / compile stand-alone.

fuzzard commented:
```Notes:
PR is not in buildable individual commits. We have broken it up like this for, hopefully, for easier review, particularly the non platform code that is touched. Before Merge, the intent is to squash most into a single commit.```

Thats a quote from the PR message. The intent is for reviewers to see non platform code at a glance for reviewers relevant areas of expertise. It will not merge with this many commits, it WILL be squashed. If you want it squashed now, thats fine by me, it will be a 3 commits.

fuzzard committed:
[TVOS] sleep state management

Implement some todo's regarding sleep/wake workflow.
Stop/start ActiveAE on sleep/wake
Stop/start All network services on sleep/wake
combine and streamline methods involved with sleep/wake

fuzzard committed:
[ios/tvos] Python3 patch update - remove waitpid define

waitpid undef not required. Functions correctly with waitpid on both ios/tvos

fuzzard commented:
Jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

kambala-decapitator commented on xbmc/platform/darwin/tvos/filesystem/TVOSDirectory.cpp:
```diff
@@ -0,0 +1,130 @@
+  return true;
+}
+
+bool CTVOSDirectory::GetDirectory(const CURL& url, CFileItemList& items)
```
auto where possible

kambala-decapitator commented on xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp:
```diff
@@ -0,0 +1,274 @@
+using namespace XFILE;
+
+
+CTVOSFile::CTVOSFile() : m_position(-1), m_pFallbackFile(nullptr)
```
set default values in class declaration, = default instead of empty body

kambala-decapitator commented on xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp:
```diff
@@ -0,0 +1,274 @@
+  if (CDarwinNSUserDefaults::GetKeyDataFromPath(m_url.Get(), lpBufInternal,
+                                                uiBufSize)) // read size of file
+  {
+    lpBufInternal = static_cast<void*>(new char[uiBufSize]);
```
use smart pointer

fuzzard replied:
discussed this one with you. At this stage i'll leave it as is. May be able to change it with some refactoring. Will look into it at a later date.

kambala-decapitator commented on xbmc/platform/darwin/tvos/filesystem/TVOSFileUtils.mm:
```diff
@@ -0,0 +1,41 @@
+const char* CTVOSFileUtils::GetUserHomeDirectory()
+{
+  static std::string appHomeFolder;
+  if (appHomeFolder.empty())
```
use std::call_once instead

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputRemote.mm:
```diff
@@ -0,0 +1,156 @@
+#import <Foundation/Foundation.h>
+#import <UIKit/UIEvent.h>
+
+@implementation TVOSLibInputRemote
```
auto where possible

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputRemote.mm:
```diff
@@ -0,0 +1,156 @@
+          static_cast<void*>(new CAction(ACTION_PLAYER_PLAYPAUSE)));
+      break;
+    case UIEventSubtypeRemoteControlPlay:
+      [g_xbmcController.inputHandler sendButtonPressed:13];
```
isn't there some enum instead of these magic numbers?

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputSettings.h:
```diff
@@ -0,0 +1,20 @@
+
+#import <Foundation/Foundation.h>
+
+extern float const REPEATED_KEYPRESS_DELAY_S;
```
1. both should have type NSTimeInterval
2. why are they defined here, but actually used in another file?

fuzzard replied:
Only reason they were thrown in here was they were some magic number hidden away in ther original location. I started to strip anything that may need to be user changable into Settings.

Happy to move out from here, as after the refactor, i dont think it ended up making sense to make it an end user setting.

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputTouch.mm:
```diff
@@ -0,0 +1,647 @@
+    case UIPressTypeRightArrow:
+      break;
+
+    default:
```
if all possible cases are covered, I suggest to remove `default`

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputTouch.mm:
```diff
@@ -0,0 +1,647 @@
+{
+  // tap side of siri remote pad
+  for (auto t : {
+         std::make_tuple(UIPressTypeUpArrow, @selector(tapUpArrowPressed:),
```
should be possible to use `{}` instead of std::make_tuple(), at least for all params after the first

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputTouch.mm:
```diff
@@ -0,0 +1,647 @@
+
+- (IBAction)handlePan:(UIPanGestureRecognizer*)sender
+{
+  if (!g_xbmcController.inputHandler.inputRemote.remoteIdleState)
```
early return for this and next if

kambala-decapitator commented on xbmc/platform/darwin/tvos/input/LibInputTouch.mm:
```diff
@@ -0,0 +1,647 @@
+                break;
+              }
+              default:
+              {
```
omit braces

kambala-decapitator commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -1258,6 +1274,10 @@ std::string CSysInfo::GetBuildTargetPlatformVersionDecoded(void)
 #elif defined(TARGET_DARWIN_IOS)
   return StringUtils::Format("version %d.%d.%d", (__IPHONE_OS_VERSION_MIN_REQUIRED / 10000) % 100,
                              (__IPHONE_OS_VERSION_MIN_REQUIRED / 100) % 100, __IPHONE_OS_VERSION_MIN_REQUIRED % 100);
+#elif defined(TARGET_DARWIN_TVOS)
```
iOS and tvOS can be merged into one using `XSTR_MACRO`

kambala-decapitator commented on xbmc/windowing/tvos/OSScreenSaverTVOS.mm:
```diff
@@ -0,0 +1,25 @@
+
+#import "platform/darwin/tvos/XBMCController.h"
+
+COSScreenSaverTVOS::COSScreenSaverTVOS()
```
= default

kambala-decapitator commented on xbmc/windowing/tvos/VideoSyncTVos.h:
```diff
@@ -0,0 +1,44 @@
+class CVideoSyncTVos : public CVideoSync, IDispResource
+{
+public:
+  CVideoSyncTVos(void* clock, CWinSystemTVOS& winSystem) : CVideoSync(clock), m_winSystem(winSystem)
```
method definition should not be in class declaration

kambala-decapitator commented on xbmc/windowing/tvos/WinSystemTVOS.mm:
```diff
@@ -0,0 +1,438 @@
+// (for ensuring that e.x. AE isn't stuck)
+constexpr uint32_t LOST_DEVICE_TIMEOUT_MS{3000};
+
+// TVOSDisplayLinkCallback is declared in the lower part of the file
```
probably defined, not declared? :)

kambala-decapitator commented on xbmc/windowing/tvos/WinSystemTVOS.mm:
```diff
@@ -0,0 +1,438 @@
+
+std::unique_ptr<CWinSystemBase> CWinSystemBase::CreateWinSystem()
+{
+  std::unique_ptr<CWinSystemBase> winSystem(new CWinSystemTVOS());
```
use `std::make_unique` instead

kambala-decapitator commented on xbmc/windowing/tvos/WinSystemTVOS.mm:
```diff
@@ -0,0 +1,438 @@
+  *h = [g_xbmcController.displayManager getScreenSize].height;
+  *fps = [g_xbmcController.displayManager getDisplayRate];
+
+  CLog::Log(LOGDEBUG, "Current resolution Screen: {} with {} x {} @  {}", screenIdx, *w, *h, *fps);
```
extra space after @

kambala-decapitator commented on xbmc/windowing/tvos/WinSystemTVOS.mm:
```diff
@@ -0,0 +1,438 @@
+void CWinSystemTVOS::FillInVideoModes(int screenIdx)
+{
+  // Potential refresh rates
+  std::vector<float> supportedDispRefreshRates = {23.976, 24.000, 25.000, 29.970,
```
add `f` suffix to all values

fuzzard commented on xbmc/platform/darwin/ios-common/IOSKeyboardView.mm:
```diff
@@ -13,100 +13,111 @@
-  {
-    [self performSelectorOnMainThread:@selector(initWithFrameInternal) withObject:nil  waitUntilDone:YES];
-  }
+  // @todo in fact, it's client's responsibility to init on main thread
```
All the keyboard stuff we were reworking. In that branch i gave you a few weeks ago regarding the ios keyboard rework. From memory, it fixed this, and removed that todo.

Was waiting on this going into master before completing the tvos part of that however

fuzzard committed:
[DarwinEmbedded] Darwinembedded generic code path for GetBuildTargetPlatformVersionDecoded

Changes to a generic code path for GetBuildTargetPlatformVersionDecoded that
utilises GetBuildTargetPlatformVersion to do the macro for the specific platform (ios
and future tvos)

fuzzard commented on xbmc/platform/darwin/tvos/TopShelf/ServiceProvider.m:
```diff
@@ -0,0 +1,121 @@
+  return TVTopShelfContentStyleSectioned;
+}
+
+- (NSArray<TVContentItem*>*)topShelfItems
```
went the __auto_type route. Was just the easier option than converting to mm

fuzzard committed:
[DarwinEmbedded] Support xcode 11 certificate naming scheme

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

chkpnt commented on docs/README.tvOS.md:
````diff
@@ -103,140 +109,141 @@ make -j$(getconf _NPROCESSORS_ONLN)
-```
-make -C tools/depends/target/binary-addons ADDONS="audioencoder.flac pvr.vdr.vnsi audiodecoder.snesapu"
-```
+**TIP:** BUILD_DIR can be omitted, and project will be created in $HOME/kodi/build
````
I just tried it out: The project has been created in `./build`.

fuzzard replied:
Not sure I see the problem?

All conventions in the readme follow on from prior steps, so if you set it up as described, Kodi source gets cloned into $HOME/Kodi, which is where you execute the cmakebuildsys command. This is then going to be ./build, as you are in the directory, or long form $HOME/Kodi/build

chkpnt replied:
You are right, I'm sorry. I've missunderstood the sentence, thought `$HOME/kodi/build` was hard coded somewere.

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

Memphiz commented:
Yay - congrats - thx and well done to all involved!

pogarek commented:
Indeed. Well done and congrats. Finally it happened !
